cpp/src/arrow/datum.cc

```diff
@@ -177,6 +177,30 @@ bool Datum::Equals(const Datum& other) const {
   }
 }
 
+Result<Datum> Datum::View(const std::shared_ptr<DataType>& out_type) const {
+  if (this->is_scalar()) {
+    std::shared_ptr<Array> tmp_array;
+    ARROW_ASSIGN_OR_RAISE(
+        tmp_array, MakeArrayFromScalar(*this->scalar().get(), 1, default_memory_pool()));
+    ARROW_ASSIGN_OR_RAISE(std::shared_ptr<ArrayData> array,
+                          ::arrow::internal::GetArrayView(tmp_array->data(), out_type));
+    ARROW_ASSIGN_OR_RAISE(std::shared_ptr<Scalar> out, MakeArray(array)->GetScalar(0));
```

Comment:
Can we instead not support this? This implementation is very heavy and isn't really "zero-copy" either.

Reply:
This was mostly to support the test which checks the invariant:

```cast_map(array[i]) == cast_map(array)[i]```

https://github.com/apache/arrow/blob/master/cpp/src/arrow/compute/kernels/test_util.cc#L127

It's not the cleanest implementation, I agree. Do you have better ideas on how it could be implemented?

Reply:
I was able to remove this function entirely by changing the early return logic.